eclipse-theia/theia
# Enable `extensions.json`

#### What it does


Fixes #8982 by adding support for `extensions.json` to the preference system.
Also fixes #9034.
Also fixes #8070 (for the @recommended case).

I'd like to see a real solution to #9255, but this PR just creates overrides to be used with this package that adds some extra rules for determining whether a preference should be saved in `extensions.json` or `settings.json`.

Here are the features of VSCode's `extensions.json` system that I'm aware of. Feel free to mention or add any I've missed:
- [x] `extensions.json` recognized by preference system.
  - [x] supports recommendations and anti-recommendations (at the moment Theia doesn't have default recommendations that need to be overridden, but downstream adopters might)
- [x] `Recommended` widget appears in plugin view.
  - [x] Installed plugins don't appear in the `Recommended` widget. 
- [x] Toast when recommended extensions detected in the workspace. 
  - [x] Preference to ignore the toast. (The VSCode preference has the same `extensions` prefix as the name of the file, which will be a bit of a challenge for our preference providers to handle, since they assume that a section name (`extensions`, `tasks`, etc.) should not be used as a prefix in the `settings` object.)
- [x] Support for `@recommended` query in Plugin View search bar
  - [x] Command to show the recommended plugins.
  - [ ] With distinction between workspace and default recommendations 

#### How to test


1. Open a workspace (single or multiroot).
2. Create an `extensions.json` file in a `.theia` or `.vscode` folder under the/a root.
3. Add some extensions inside the `recommendations` array of the following structure: `{"recommendations": []}`
4. Observe that the schema warns if your extensions don't follow the format `<alphanumeric or - >.<alphanumeric or - >` (with `-` barred from first position in both alphanumeric groups)
5. Open the plugins view. 
6. Observe that a widget with the heading `Recommended` is present. (May require resetting workbench layout.)
7. Observe that if you add any real plugins to the JSON, they appear in that widget, unless they're already installed (you can check in the `Installed` widget.
8. Of the plugins you've added to `recommendations`, add a few to a new array labeled `unwantedRecommendations` (leaving them in `recommendations`, as well). Observe that any listed under `unwantedRecommendations` do not appear in the `Recommended` widget.
9. You should be able to add an `extensions.json` file to each root of a multi-root workspace, and the `Recommended` widget should aggregate and display (all recommendations) - (all anti-recommendations). That is, an anti-recommendation anywhere will override a recommendation anywhere, even if they aren't the same folder.
10. It should be possible to add an `extensions` object with the same schema to the `settings` object of a workspace file (after #8917, it will be possible to put `extensions` outside the `settings` object.)
11. Open the preferences UI and modify the value of `extensions.ignoreRecommendations`. Observe that the value is set in a `settings.json` file, not an `extensions.json` file.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
